### PR TITLE
Property-based tests for UTxO statistics

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#549 

# Overview

- [x] I have removed some dead-code (`UTxOStatisticsError`, curious that weeder didn't caught that...)
- [x] I have added `Buildable` instance to `HistogramBar` to display richer errors in QC failures
- [x] I have added 4 property-based tests as suggested in the original ticket.

# Comments

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs [hunk 1]
```diff
@@ -302,6 +323,63 @@ prop_2_6_2 (ins, u) =
             ===
         balance u - balance (u `restrictedBy` ins)
 
+{-------------------------------------------------------------------------------
+                        UTxO statistics Properties
+-------------------------------------------------------------------------------}
+
+-- | The 'total' stake in the statistics is the UTxO's balance
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:330]: So just that the "total" (API terminology) field is the "sum" of the UTxO, right.

Is "stake" the correct term here?

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs [hunk 2]
```diff
@@ -302,6 +323,63 @@ prop_2_6_2 (ins, u) =
...
+    :: BoundType
+    -> ShowFmt UTxO
+    -> Property
+propUtxoTotalIsBalance bType (ShowFmt utxo) =
+    fromIntegral totalStake == balance utxo
+    & cover 75 (utxo /= mempty) "UTxO /= empty"
+  where
+    UTxOStatistics _ totalStake _ = computeUtxoStatistics bType utxo
+
+-- | The sum of the weighted distribution is greater than the sum of UTxOs
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:341]: > weighted distribution is greater than the sum of UTxOs

I don't get this at all… Why is this a **weighted** distribution? If it is, it's not relevant here?

But ok, the bucket "10" really contains "1-10", so multiplying buckets values with counts should `>=`.

**Review by piotr-iohk — APPROVED** [review]
¯\_(ツ)_/¯... :+1: 

(I get the idea of what is tested I think, code-wise I trust it's OK)

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs [hunk 3]
```diff
@@ -302,6 +323,63 @@ prop_2_6_2 (ins, u) =
...
+    -> Property
+propUtxoEmptyIsEmpty bType (ShowFmt utxo) =
+    if all isEmpty bars then utxo === mempty else utxo =/= mempty
+    & cover 75 (utxo /= mempty) "UTxO /= empty"
+    & counterexample ("Histogram: " <> pretty bars)
+  where
+    UTxOStatistics bars _ _ = computeUtxoStatistics bType utxo
+    isEmpty (HistogramBar _ v) = v == 0
+
+-- | The sum of the distribution coefficients should is equal to the number of
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:369]: > distribution coefficients

This also radically confused me. It appears to be _a thing_, but googling and finding _is the ratio of concentrations of a compound in a mixture of two immiscible phases _ doesn't help me understand this.

I'm not sure what the optimal term is, but maybe something more like "count" or "frequency", or "the sum of the frequencies of all buckets/bins" 🤷‍♂ 

But this is at least the only place where bin-counts/whatever are called "coefficients"…

Or am I just missing how one should view this?

**Review by Anviking — APPROVED** [review]
Looks great!

But I found myself really confused by a few choices of terminology in the comments, and took a while to figure this out being completely foreight to the UTXO Statistics functionality & this code.

So… a few comments/questions/potential suggestions, thinking out loudly a bit 😅

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:330]: I think it is :)

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:341]: Each value is in a "bucket". So, if we computed a weighted sum where the weights are the bucket upper-bound, we get a value that should be bigger than the original value distributed over all buckets. That's what called a "weighted distribution" in short :sweat_smile:

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:369]: > <strong>distribution</strong>  
> The action of sharing something out among a number of recipients.

---

> <strong>coefficient</strong>  
> A numerical or constant quantity placed before and multiplying the variable in an algebraic expression (e.g. 4 in 4x y).


Distributions are somewhat "well-known" in mathematics: https://en.wikipedia.org/wiki/Distribution_(mathematics) hence why I coined this term here. Especially in the context of an histogram:

> <strong>histogram</strong>  
> A histogram is an accurate representation of the distribution of numerical data.

Perhaps _bucket's coefficient_ works fine too, like _bucket's count_ or _bucket's frequency_ but they're all really synonym to me :no_mouth:

File: lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs (hunk 1 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:330]: What if the UTxO contains funds within enterprise addresses (i.e which are non-stakeable)?

IMG_EB54401351E6-1

(https://hydra.iohk.io/build/790053/download/1/delegation_design_spec.pdf)

**Pull request merged**